lib/galaxy/model/__init__.py

```diff
@@ -7546,29 +7554,29 @@ def dataset_instances(self):
 
     @property
     def has_deferred_data(self):
-        return self.element_object.has_deferred_data
+        return self.element_object and self.element_object.has_deferred_data
 
     def copy_to_collection(
         self,
-        collection,
-        destination=None,
-        element_destination=None,
-        dataset_instance_attributes=None,
+        collection: DatasetCollection,
+        destination: Optional[HistoryDatasetCollectionAssociation] = None,
+        element_destination: Optional[History] = None,
+        dataset_instance_attributes: Optional[Dict[str, Any]] = None,
         flush=True,
         minimize_copies=False,
     ):
         dataset_instance_attributes = dataset_instance_attributes or {}
         element_object = self.element_object
         if element_destination:
-            if self.is_collection:
+            if isinstance(element_object, DatasetCollection):
                 element_object = element_object.copy(
                     destination=destination,
                     element_destination=element_destination,
                     dataset_instance_attributes=dataset_instance_attributes,
                     flush=flush,
                     minimize_copies=minimize_copies,
                 )
-            else:
+            elif isinstance(element_object, HistoryDatasetAssociation):
```

Comment:
Should we raise an error in case `element_object` is a `LibraryDatasetDatasetAssociation` ?

Reply:
Yeah, that makes sense.

Reply:
Done in https://github.com/galaxyproject/galaxy/pull/19961/commits/8b3a75e2b8cc095e9e3026350f13661689332a71